src/util/futurepool.rs

```diff
@@ -0,0 +1,355 @@
+            delegator
+        } else {
+            panic!("Called from threads out of the future thread pool");
+        }
+    }
+}
+
+/// A future thread pool that supports `on_tick` for each thread.
+#[derive(Debug)]
+pub struct FuturePool<T: Context + 'static> {
+    pool: CpuPool,
+    context_delegators: ContextDelegators<T>,
+    running_task_count: Arc<AtomicUsize>,
+}
+
+impl<T: Context + 'static> Clone for FuturePool<T> {
+    fn clone(&self) -> FuturePool<T> {
+        FuturePool::<T> {
+            pool: self.pool.clone(),
+            context_delegators: self.context_delegators.clone(),
+            running_task_count: Arc::clone(&self.running_task_count),
+        }
+    }
+}
+
+impl<T: Context + 'static> util::AssertSend for FuturePool<T> {}
+impl<T: Context + 'static> util::AssertSync for FuturePool<T> {}
+
+impl<T: Context + 'static> FuturePool<T> {
+    pub fn new<F>(
+        pool_size: usize,
+        stack_size: usize,
+        name_prefix: &str,
+        tick_interval: Duration,
+        context_factory: F,
+    ) -> FuturePool<T>
+    where
+        F: Send + 'static + Fn() -> T,
+    {
+        let (tx, rx) = mpsc::sync_channel(pool_size);
+        let pool = cpupool::Builder::new()
+            .pool_size(pool_size)
+            .stack_size(stack_size)
+            .name_prefix(name_prefix)
+            .after_start(move || {
+                // We only need to know each thread's id and we can build context later
+                // by invoking `context_factory` in a !Sync way.
+                let thread_id = thread::current().id();
+                tx.send(thread_id).unwrap();
+            })
+            .create();
+        let contexts = (0..pool_size)
+            .map(|_| {
+                let thread_id = rx.recv().unwrap();
+                let context = context_factory();
+                let context_delegator = ContextDelegator::new(context, tick_interval);
+                (thread_id, context_delegator)
+            })
+            .collect();
+        FuturePool {
+            pool,
+            context_delegators: ContextDelegators::new(contexts),
+            running_task_count: Arc::new(AtomicUsize::new(0)),
+        }
+    }
+
+    /// Get current running task count
+    #[inline]
+    pub fn get_running_task_count(&self) -> usize {
+        self.running_task_count.load(Ordering::Acquire)
+    }
+
+    pub fn spawn<F, R>(&self, future_factory: R) -> CpuFuture<F::Item, F::Error>
+    where
+        R: FnOnce(ContextDelegators<T>) -> F + Send + 'static,
+        F: Future + Send + 'static,
+        F::Item: Send + 'static,
+        F::Error: Send + 'static,
+    {
+        let running_task_count = Arc::clone(&self.running_task_count);
+        let delegators = self.context_delegators.clone();
+        let func = move || {
+            future_factory(delegators.clone()).then(move |r| {
+                let delegator = delegators.get_current_thread_delegator();
+                delegator.on_task_finish();
```

Comment:
Is it possible the some thread's context never got flushed?

Reply:
Maybe. As long as there are no tasks being executed on the thread, the context's `on_tick` will not be triggered. I and @hicqu thinks current triggering guarantee provided by this futurepool is enough for now.

Reply:
If we don't care accuracy, I think it is acceptable here unless we think a better way.